traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
-        A weak reference to a HasTraits instance that the list has been set
-        as an attribute for.
-    name : str
-        The name of the attribute on the object.
-    name_items : str
-        The name of the items event trait that the trait list will fire when
-        mutated.
+    value : list
+        The value for the list
+    validator : callable
+        Called to validate items in the list
+    notifiers : list of callable
+        A list of callables with the signature::
+
+         notifier(trait_list, index, removed, added)
+
     """
 
-    def __init__(self, trait, object, name, value):
-        self.trait = trait
-        self.object = ref(object)
-        self.name = name
-        self.name_items = None
-        if trait.has_items:
-            self.name_items = name + "_items"
+    # ------------------------------------------------------------------------
+    # TraitList interface
+    # ------------------------------------------------------------------------
+
+    def validate(self, index, removed, added):
+        """ Validate values for given index and removed values.
 
-        # Do the validated 'setslice' assignment without raising an
-        # 'items_changed' event:
-        if trait.minlen <= len(value) <= trait.maxlen:
-            try:
-                validate = trait.item_trait.handler.validate
-                if validate is not None:
-                    value = [validate(object, name, val) for val in value]
+        This simply calls the validator provided by the class, if any.
+        The validator is expected to have the signature::
 
-                list.__setitem__(self, slice(0, 0), value)
+            validator(trait_list, index, removed, added)
 
-                return
+        and return a validated added value or list of added values
+        or raise a TraitError.
 
-            except TraitError as excp:
-                excp.set_prefix("Each element of the")
-                raise excp
+        Parameters
+        ----------
+        index : int or slice
+            The indices being modified by the operation.
+        removed : list
+            The items to be removed.
+        added : list
+            The new items being added to the list.
 
-        self.len_error(len(value))
+        Returns
+        -------
+        values : list
+            The validated values that must be added to the list.
 
-    def _send_trait_items_event(self, name, event, items_event=None):
-        """ Send a TraitListEvent to the owning object if there is one.
+        Raises
+        ------
+        TraitError : Exception
+            If validation fails.
         """
-        object = self.object()
-        if object is not None:
-            if items_event is None and hasattr(self, "trait"):
-                items_event = self.trait.items_event()
-            object.trait_items_event(name, event, items_event)
+
+        # Use getattr as pickle can call `extend` before validator is set.
```

Comment:
Would this problem go away if we weren't validating at `__init__` time? The code below seems a bit hacky, and seems as though it's probably in the wrong place.

The comment is potentially confusing: it's not clear to the code reader why `pickle` would be relevant here. I suggest expanding the explanation a bit.

Reply:
Not used anymore .